File: 2024 1월 3주차/백준 1365.py
def binary_search(sub, val):
    left, right = 0, len(sub) - 1
    while left <= right:
        mid = left + (right - left) // 2
        if sub[mid] < val:
            left = mid + 1
        else:
            right = mid - 1
    return left

def min_cut(connections):
    sub = []
    for val in connections:
        pos = binary_search(sub, val)
        if pos == len(sub):
            sub.append(val)
        else:
            sub[pos] = val
    return len(connections) - len(sub)

N = int(input())
connections = list(map(int, input().split()))
print(min_cut(connections))

File: 2024 1월 3주차/백준 1365.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');

const N = parseInt(input[0]);
const connections = input[1].split(' ').map(Number);

function binarySearch(sub, val) {
    let left = 0, right = sub.length - 1;
    while (left <= right) {
        let mid = Math.floor((left + right) / 2);
        if (sub[mid] < val) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return left;
}

function minCut(connections) {
    let sub = [];
    for (let val of connections) {
        let pos = binarySearch(sub, val);
        if (pos === sub.length) {
            sub.push(val);
        } else {
            sub[pos] = val;
        }
    }
    return connections.length - sub.length;
}

console.log(minCut(connections));